### ISSUE TITLE
Should we check invariants in constructors to simplify the code?

### ISSUE THREAD

**Issue by seanmcl** [typelevel/paiges#96]
If we trusted the stated invariants, such as

https://github.com/typelevel/paiges/blob/master/core/src/main/scala/org/typelevel/paiges/Doc.scala#L585

we could simplify the code, e.g.

https://github.com/typelevel/paiges/blob/master/core/src/main/scala/org/typelevel/paiges/Doc.scala#L192

https://github.com/typelevel/paiges/blob/master/core/src/main/scala/org/typelevel/paiges/Doc.scala#L185

How about we either check the invariants in the constructors, or if performance is a concern, just let the code above fail?  There is a cost in conceptual clarity to being defensive.

I'm partly asking to help understand the culture of the project.  There are some idioms here with which I'm unfamiliar.

Thanks!

**Comment by johnynek** [comment]
I would say correctness is number 1, achieving it via types over runtime checks is number 2, performance is number 3.

In those examples, which aren’t hit in every use, we do check something we could remove, but empty check on a String is very fast.

If we put a check in every allocation of Text we would pay everywhere for it.

An alternate idea might be to make an AnyVal class for NonEmptyString. Then we could probably make them by construction in a few places without having to do extra work.

Of course, Text may basically be NonEmptyString. So maybe we can make the constructor private and control where we create them.

Basically, I’m nervous about checking in every allocation but maybe it would actually be cheaper. This is not a data driver position at this moment.

**Comment by seanmcl** [comment]
This issue was partly a misunderstanding on my part that the constructors aren't exposed and the factory constructors enforce the invariants.  It still means the implementers can make mistakes, but it seems pretty good as it is.  I do vote for removing the code that tries to correct for possibly broken invariants like those above.  I think it makes the code harder to understand.  I'd prefer if we changed those to assertions that raise a RuntimeException.  But I'm happy to close this issue for now.
